driver-core/src/test/functional/com/mongodb/client/model/expressions/StringExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,156 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import org.junit.jupiter.api.Test;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SuppressWarnings({"ConstantConditions"})
+class StringExpressionsFunctionalTest extends AbstractExpressionsFunctionalTest {
+    // https://www.mongodb.com/docs/manual/reference/operator/aggregation/#string-expression-operators
+
+    private final String sushi = "\u5BFF\u53F8";
+    private final String jalapeno = "jalape\u00F1o";
+
+    @Test
+    public void literalsTest() {
+        assertExpression("", of(""), "''");
+        assertExpression("abc", of("abc"), "'abc'");
+        assertThrows(IllegalArgumentException.class, () -> of((String) null));
+        assertExpression(sushi, of(sushi), "'" + sushi + "'");
+    }
+
+    @Test
+    public void concatTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/concat/
+        assertExpression(
+                "abc".concat("de"),
+                of("abc").concat(of("de")),
+                "{'$concat': ['abc', 'de']}");
+    }
+
+    @Test
+    public void toLowerTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/toLower/
+        assertExpression(
+                "ABC".toLowerCase(),
+                of("ABC").toLower(),
+                "{'$toLower': 'ABC'}");
+    }
+
+    @Test
+    public void toUpperTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/toUpper/ (?)
+        assertExpression(
+                "abc".toUpperCase(),
+                of("abc").toUpper(),
+                "{'$toUpper': 'abc'}");
+    }
+
+    @Test
+    public void strLenTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/strLenCP/ (?)
+        // TODO naming: just strLen, lengthCP, lengthCodePoints, stringLengthInCodePoints?...
+        assertExpression(
+                "abc".length(),
+                of("abc").strLen(),
+                "{'$strLenCP': 'abc'}");
+
+        // unicode
+        assertExpression(
+                jalapeno.length(),
+                of(jalapeno).strLen(),
+                "{'$strLenCP': '" + jalapeno + "'}");
+        assertExpression(
+                sushi.length(),
```

Comment:
We should replace all `String.length` invocations in these tests with `String.codePointCount`. For example, `sushi.codePointCount(0, suchi.length())`.

The tests currently pass just by a chance, and you can see `assertExpression(sushi.length(), of(sushi).strLen(), "{'$strLenCP': '" + sushi + "'}")` failing if instead of `sushi = "\u5BFF\u53F8"` (`寿司`)  you use `sushi = "\u5BFF\u53F8\uD83E\uDD84"` (`寿司🦄`), where the third code point corresponding to the character `🦄` is encoded as two UTF-16 16-bit code units (`\uD83E` and `\uDD84`).